gubernator/github_auth_test.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+
+# Copyright 2016 The Kubernetes Authors All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+import urlparse
+
+import webtest
+
+import gcs_async_test
+import main
+
+main.app.config['github_client'] = {
+    'id': '12345',
+    'secret': 'swordfish',
+}
+
+app = webtest.TestApp(main.app)
+
+VEND_URL = 'https://github.com/login/oauth/access_token'
+USER_URL = 'https://api.github.com/user'
+
+class TestGithubAuth(unittest.TestCase):
+    def setUp(self):
+        app.reset()
+        self.testbed.init_app_identity_stub()
+        self.testbed.init_urlfetch_stub()
+        self.calls = []
+        self.results = {
+            VEND_URL: ('{"access_token": "token"}', 200),
+            USER_URL: ('{"login": "foo"}', 200),
+        }
+        gcs_async_test.install_handler_dispatcher(
+            self.testbed.get_stub('urlfetch'),
+            (lambda url: url in self.results),
+            self.dispatcher)
+
+    def dispatcher(self, method, url, payload, headers):
+        self.calls.append([method, url, payload, headers])
+        return self.results[url]
+
+    @staticmethod
+    def do_phase1(arg=''):
+        return app.get('/github_auth' + arg)
+
+    @staticmethod
+    def parse_phase1(phase1):
+        parsed = urlparse.urlparse(phase1.location)
+        query = urlparse.parse_qs(parsed.query)
+        state = query.pop('state')[0]
+        return state, query
+
+    def do_phase2(self, phase1=None, status=None):
+        if not phase1:
+            phase1 = self.do_phase1()
+        state, query = self.parse_phase1(phase1)
+        code = 'somerandomcode'
```

Comment:
Since you later check this value in test_login_works can you make this a class variable or module constant?